spinalcordtoolbox/image.py

```diff
@@ -503,53 +503,44 @@ def save(self, path=None, dtype=None, verbose=1, mutable=False):
 
         :param mutable: whether to update members with newly created path or dtype
         """
+        if not mutable:
+            self.copy().save(path, dtype, verbose, mutable=True)
+            return self
 
-        if path is None and self.absolutepath is None:
-            raise RuntimeError("Don't know where to save the image (no absolutepath or path parameter)")
-        elif path is not None and os.path.isdir(path) and self.absolutepath is not None:
+        if path is None:
+            # Use self.absolutepath by default
+            if self.absolutepath is None:
+                raise ValueError("Don't know where to save the image (no absolutepath or path parameter)")
+            path = self.absolutepath
+        elif os.path.isdir(path):
             # Save to destination directory with original basename
+            if self.absolutepath is None:
+                raise ValueError("Don't know where to save the image (no absolutepath and path parameter is dir)")
             path = os.path.join(os.path.abspath(path), os.path.basename(self.absolutepath))
```

Comment:
I had a little trouble understanding these `path` changes, too. (I couldn't easily tell what all of the cases were?)

This is mostly personal preference, but... I wanted to see if I could clarify each case?

- I changed the conditionals to check the positive case (`if self.absolutepath`)
- I added `else:` statements to give each case the same indentation
- I moved the comments around a little for visual clarity

```python
       # Case 1: `path` not specified
       if path is None:
            if self.absolutepath:  # Fallback to the original filepath
                path = self.absolutepath  
            else:
                raise ValueError("Don't know where to save the image (no absolutepath or path parameter)")
        # Case 2: `path` points to an existing directory
        elif os.path.isdir(path):
            if self.absolutepath:  # Use the original filename, but save to the directory specified by `path`
                path = os.path.join(os.path.abspath(path), os.path.basename(self.absolutepath))
            else:
                raise ValueError("Don't know where to save the image (path parameter is dir, but absolutepath is missing)")
        # Case 3: `path` points to a file
        elif :
            self.absolutepath = path
```

Reply:
Case 3 isn't really a case, actually: the result of the `if path is None` and `elif os.path.isdir(path)` cases is to make sure that the `path` variable is set to the actual filename we want to use, in case it's not already usable. After that, the `self.absolutepath = path` assignment should be unconditional: regardless of how we got the final value of `path`, we need to record it as this image's filename.

If you want to keep the last `else` for clarity, you could replace it with either `pass` or `path = path` (with a code comment in either case). Maybe an extra blank line before the `self.absolutepath = path` and/or a code comment as well?

Reply:
(But overall this is a fine suggestion, feel free to commit it yourself.)

Reply:
There is one corner case we're not checking for, though: 

`os.path.isdir(path)` checks if paths are [_existing_](https://docs.python.org/3/library/os.path.html#os.path.isdir) directories or not. Meaning, the user could erroneously pass a nonexistent directory, "Case 2" would be skipped, our code would presume that this directory is actually a file (Case 3), and `nib.save()` would likely fail.

So, it might be beneficial to distinguish between "nonexistent directory" and "file"?

```python
       # Case 1: `path` not specified
       if path is None:
            if self.absolutepath:  # Fallback to the original filepath
                path = self.absolutepath  
            else:
                raise ValueError("Don't know where to save the image (no absolutepath or path parameter)")
        # Case 2: `path` points to an existing directory
        elif os.path.isdir(path):
            if self.absolutepath:  # Use the original filename, but save to the directory specified by `path`
                path = os.path.join(os.path.abspath(path), os.path.basename(self.absolutepath))
            else:
                raise ValueError("Don't know where to save the image (path parameter is dir, but absolutepath is missing)")
        # Case 3: `path` has no extension, so it is a *nonexistent* directory
        elif os.path.splitext(path)[1] == '':
            if self.absolutepath:  # Use the original filename, but save to the directory specified by `path`
                os.mkdir(os.path.abspath(path))
                path = os.path.join(os.path.abspath(path), os.path.basename(self.absolutepath))
            else:
                raise ValueError("Don't know where to save the image (path parameter is dir, but absolutepath is missing)")     
        # Case 4: `path` points to a file
        else:
            self.absolutepath = path
```

Maybe checking the extension is a bit faulty, though? 

(e.g. What happens if, say, we're doing some filename manipulation, and we accidentally forget to append the extension? My proposed approach will quietly create a new directory and save the file using the absolutepath without error.)

So, maybe we _want_ `nib` to throw the error, rather than trying to catch it ourselves...

Reply:
Ahhh, of course! Case 3 isn't a case, you're right. Thanks for the correction there!

Reply:
> (But overall this is a fine suggestion, feel free to commit it yourself.)

Done in[22a24a2](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3944/commits/22a24a26e8831d06599318e4e135bef5e6c02f7e). :)

Reply:
I think trying to guess whether a path is a nonexistent file or a nonexistent directory based on extension is more likely to mask errors than to be helpful, yes. (To be honest I'm surprised we even have special logic for the existing directory case.)